yolo_to_yolo/data_transformers.py

```diff
@@ -47,3 +47,137 @@ def __call__(self, input_data: YoloImageData) -> Iterable[YoloImageData]:
+        self.min_size = min_size
+        self.min_padding = min_padding
+        self.min_char_overlap = min_char_overlap
+
+    def __call__(self, input_data: YoloImageData) -> Iterable[YoloImageData]:
+        img_height, img_width = input_data.image.shape[:2]
+        
+        # I hope this is right
+        shape_labels = [label for label in input_data.labels if "shape" in label.classname]
+        char_labels = [label for label in input_data.labels if "character" in label.classname]
+        
+        for shape_label in shape_labels:
+            if isinstance(shape_label.location, YoloBbox):
+                shape_bbox = shape_label.location
+                best_char_label = None
+                best_overlap = self.min_char_overlap
+
+                # Find best match for character bbox (basically the one that overlaps the most with the shape bbox)
+                for char_label in char_labels:
+                    if isinstance(char_label.location, YoloBbox):
+                        char_bbox = char_label.location
+                        overlap = self._calculate_iou(shape_bbox, char_bbox)
+
+                        if overlap > best_overlap:
+                            best_overlap = overlap
+                            best_char_label = char_label
+
+                x1, y1, x2, y2 = self._get_combined_bbox(shape_bbox, best_char_label, img_width, img_height)
+
+                # print(f"Combined BBox for {shape_label.classname}: ({x1}, {y1}) to ({x2}, {y2})")
+
+                # Apply minimum size and padding
+                width = max(x2 - x1, self.min_size[0])
+                height = max(y2 - y1, self.min_size[1])
+                x1 = max(x1 - self.min_padding, 0)
+                y1 = max(y1 - self.min_padding, 0)
+                x2 = min(x1 + width + self.min_padding * 2, img_width)
+                y2 = min(y1 + height + self.min_padding * 2, img_height)
+
+                # Apply crop
+                cropped_image = input_data.image[y1:y2, x1:x2]
+
+                # print(f"Cropped Image Shape for {shape_label.classname}: {cropped_image.shape}")
+
+                # Create new labels for the cropped image
+                new_labels = [YoloLabel(location=YoloBbox(x=0.5, y=0.5, w=1.0, h=1.0), classname=shape_label.classname)]
+                if best_char_label:
+                    new_labels.append(YoloLabel(location=YoloBbox(x=0.5, y=0.5, w=1.0, h=1.0), classname=best_char_label.classname))
+
+                # Make new image data to yield
+                new_img_data = YoloImageData(
+                    img_id=f"{input_data.img_id}_{shape_label.classname}",
+                    task=input_data.task,
+                    image=cropped_image,
+                    labels=new_labels
+                )
+
+                yield new_img_data
+
+    def _calculate_iou(self, bbox1: YoloBbox, bbox2: YoloBbox) -> float:
```

Comment:
Slight nitpick, but methods that don't rely on instance state should be marked `staticmethod`.